### ISSUE TITLE
Using @Controller with @ContentType throws me an error

### ISSUE THREAD

**Issue by appsolutegeek** [pleerock/routing-controllers#236]
Hi,

I have managed to get my endpoints working using @JsonController, works flawlessly. I have now changed over to @Controller, the reason being is that one of the endpoints needs to be something other than application/json.

So I have decorated my method (Its a POST) with 

       @ContentType("application/json"

Using postman I initiate a call but receive "Unexpected 'E'"

and the node session crashes with the following

> Error
>     at BadRequestError.HttpError [as constructor]

Removing the ContentType decorator and changing back to JsonController works.

I can confirm I am calling in using Content-Type: application/json 

Anyone know whats going on ?

**Comment by 19majkel94** [comment]
`@ContentType("application/json")` is only setting response header AFAIK. So the returned object isn't transformed to json but to plain string. So when the error is throwing, the response body is not a JSON with name, message and stack properties but a text representation just like it's printed on the console.

@pleerock @NoNameProvided 
I think we have two options:
1. Put content-type header only after sucessfull action call.
1. Detect JSON content type and mark the action as JSONtyped here:
https://github.com/pleerock/routing-controllers/blob/master/src/metadata/ActionMetadata.ts#L201
And if regex is bad, we should introduce back again `@TextResponse` and `@JsonResponse`.

And I think that now we have a big problem with mixed controllers - JSON endpoint with one text reponse (like server side rendering react app or generating svg in my case). `@TextResponse` is gone, right? So how to handle now this kind of cases?

**Comment by NoNameProvided** [comment]
I dont like the idea of having `@TextResponse` and `@JsonResponse`. Users may want to send a different type of contents down. 

Instead we should simply treat `@ContentType("application/json")` specially as you wrote in your second point if I understand you correctly. So we should automatically cast objects to json when the `application/json` content type has been set.

**Comment by 19majkel94** [comment]
So we should check if the content type is like `/application\/json/` and mark the action as json typed, so it will have json formated errors and response (`res.json(result)`). I other cases we just use `send()` and place correct type header to response, right?

**Comment by NoNameProvided** [comment]
Yes, I think that is the best approach.

**Comment by 19majkel94** [comment]
I would still prefer only `@JsonResponse` as a replacement for `@ContentType("application/json")` to mark the action as JSONtyped and set correct header. For other content types we should still use `@ContentType` decorator, use `send()` and set header.

**Comment by NoNameProvided** [comment]
I find that confusing, if we have `@JsonResponse` why we don't have `@TextResponse` or `@BinaryResponse`, etc? If someone uses the `@Controller` decorator, then I think it's better to have a unified way of setting the headers and not having any special decorator.

**Comment by 19majkel94** [comment]
Because JSON action is a special case. We have to decide how we will return errors for text/binary/json responses because successful response has no problem. So with JSON action we use different method (`json()` instead of `send()`) and also we have to format error response as JSON. All the other cases (octet/html/plain) are only about headers with no different treatment on framework level.

**Comment by NoNameProvided** [comment]
Ok, I have changed my mind, so actually as we have `@JsonController` it's ok to have `@JsonResponse` as well. To treat JSON as a first-class citizen in routuing-controllers.

However we have to cover that case if the user sets the `@ContentType` to json but not the `@ JsonResponse ` decorator.

**Comment by 19majkel94** [comment]
`@ContentType` should be treated only to set the response header, `@JsonResponse` to also mark an action as JSONtyped along with `@JsonController`. We should print error on console when content-type is json but not using `@ JsonResponse`.

We should also detect the situation where we have `@JsonController` and `@ContentType("text/html") on action - we should unmark the action as JSONtyped.

**Comment by d-bechtel** [comment]
What about the intention of the issue creator? I think he want to reject requests which don't have the correct MediaType. What about a new decorator @Consumes('application/json'). If Content-Type header is not correct, a bad request error will be returned. The @ContentType decorator should be renamed to @Produces

**Comment by 19majkel94** [comment]
>  I think he want to reject requests which don't have the correct MediaType.

Could you quote the part of his text when he mention this? He reported the bug as we discussed here.

**Comment by d-bechtel** [comment]
ok, reread it again.. maybe i'm wrong but i think my proposal is a good feature request

**Comment by 19majkel94** [comment]
@d-bechtel So please create a separate issue where you describe your proposal in more details.

**Comment by pleerock** [comment]
I dont like idea of bringing `@JsonResponse` back. Agree with @NoNameProvided on point of having all decorators then. But better not to have those decorators at all.

Also I remember I wanted to remove `@JsonController` and left only `@Controller` and automatically determine what to send - json or text based on returned data. But I failed to implement it, faced some problems (don't remember what problems).

And since I don't remember what problems I had - I ask myself question again - can we return data based on returned content? If object is returned - "json", if string - "text", etc.?

**Comment by 19majkel94** [comment]
`null` and string are valid JSON values, so some users may encounter problems when they return string in their json rest api but it will have content-type header set to text/plain (or html?).

We can get rid of `@JsonController` by declaring `@ContentType` on controller level, which can be overwritten by controller's method decorator for selected action.

**Comment by pleerock** [comment]
I remember one problem I had - is what to call "json" or "text" if nothing is returned, like in delete action - we still need to return "application/json" header even if nothing is returned.

**Comment by pleerock** [comment]
So what is final decision then? For me it looks like its better to check using regexp then introduce a that ugly decorator which introduce more confusion and need users knowledge about this decorator.

**Comment by NoNameProvided** [comment]
I can accept both version. In one hand the best way would be to transform based on the content-type set by the `@ ContentType` decorator. 

In the other hand we already have `@JsonController` so it's ok for me to bring a `@JsonResponse` decorator as well.

**Comment by 19majkel94** [comment]
> In the other hand we already have @JsonController so it's ok for me to bring a @JsonResponse decorator as well.

The problem is with returning html in json controller - endpoint rendering eg. html report, pdf report or xls doc.

**Comment by pleerock** [comment]
lets just do a first approach then

**Comment by 19majkel94** [comment]
So the PR should remove `@JsonController` in favour of `@ContentType("application/json")` on controller level which can be overridden by `@ContentType` on action level? So the `isJsonTyped` would be determined dynamically per route. `@ContentType("text/html")` overriding `@JsonController` is not intuitive. Also we should provide enums for basic content type, like html, text, json, etc.

**Comment by pleerock** [comment]
wait wait we didnt talk about `@JsonController` removal, right? I don't think we can do it easily without bringing even bad design, right? I was talking to apply a simple regexp to check if content type is set to json and use `json` method instead of `send` as you suggested before
